### ISSUE TITLE
Clarify what "Gradle installation directory" means for `gradlew`

### ISSUE THREAD

**Issue by joffrey-bion** [gradle/gradle#26088]
### Issue type

Missing information

### Problem description

In the [list for the order of precedence](https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_configuration_properties) of Gradle properties, the last location is specified as the `Gradle installation directory`.

What does it mean when using the Gradle wrapper mechanism in a project? Is there such a location at all in that case? It would be nice for this information to be present in the doc, as I couldn't find it so far.

### Page with the problem

https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_configuration_properties

**Comment by joffrey-bion** [comment]
By the way, if there is no installation directory for the Gradle wrapper, then how can I specify a default value for the whole machine for a given property (say, `org.gradle.java.installations.fromEnv`) that is still overridable on per-project basis? (it must not have precedence over the project's decision in the project's `gradle.properties`).

This would be a typical use case for CI runners that provide multiple JDK installations, looping back to:
https://github.com/gradle/gradle/issues/14903#issuecomment-1674452300

**Comment by asteiner-psh** [comment]
Looking at the PR that closed this ticket, I do not see any changes that addresses the reported issue (Clarify what "Gradle installation directory" means for gradlew)

Also, in the PR, this issue is mentioned with an incorrect label: "Fix build order preference - https://github.com/gradle/gradle/issues/26088"

Is it possible this issue was incorrectly referenced in that PR?

**Comment by lkasso** [comment]
Hey @asteiner-psh, the changes aren't public yet.

I updated the two files with this question in mind (note this is still not released on https://docs.gradle.org/current):
- https://docs.gradle.org/nightly/userguide/gradle_wrapper.html
- https://docs.gradle.org/nightly/userguide/build_environment.html?

Gradle and the Gradle wrapper use the same order preference.

Is there something I did not answer for you and @joffrey-bion? Please let me know so I can make further updates.

**Comment by joffrey-bion** [comment]
From the new formulation, it seems that we don't actually need a Gradle installation in the directory that `GRADLE_HOME` points to in order to define properties. Is this true?

If I create a `/gradle_defaults/gradle.properties` file with some properties, and set `GRADLE_HOME=/gradle_defaults` (where `/gradle_defaults` does *not* contain a gradle installation), will it be used in prio order 3.2 by projects using the wrapper?

If yes, this mean I could use it to define default properties that can be overriden on a per-project basis.

**Comment by lkasso** [comment]
@joffrey-bion, Gradle will automatically store all downloaded versions of Gradle in `GRADLE_USER_HOME`. That way, as you switch Gradle versions in your projects, you don't have to download a fresh copy.
This is not to be confused with `GRADLE_HOME`. 
The `GRADLE_HOME` is entirely optionally and rarely used thanks to the Gradle Wrapper. 
When `GRADLE_HOME` is used, it is to specify a local Gradle version. This is often used by Gradle developers testing a local custom Gradle copy.

What you want to do is put your gradle.properties file in `GRADLE_USER_HOME`.

**Comment by joffrey-bion** [comment]
> What you want to do is put your gradle.properties file in GRADLE_USER_HOME, not GRADLE_HOME.

No, `GRADLE_USER_HOME` doesn't work, because properties defined there will override properties from the project's `gradle.properties`. 

Again, I want to specify the default properties for my machine, but still let the project's `gradle.properties` override them as needed.

The only place to do that in the list is the `GRADLE_HOME`, not the `GRADLE_USER_HOME`.

> Gradle and the Wrapper don't use GRADLE_HOME.

What do you mean? Then why is this part of the list of places where `gradle.properties` are searched?

**Comment by joffrey-bion** [comment]
Also, I think it would be rather important not to conflate the types of properties and the mechanisms to set them. It is irrelevant to consider the precedence between 2 different properties. We should see 3 independent tables of priorities: one for system properties, one for gradle properties, one for project properties. And in each table, list all mechanisms to set such properties (which will be similar but not the same).

I expanded much more in the `Precedence` section in this issue: https://github.com/gradle/gradle/issues/26107

**Comment by lkasso** [comment]
@joffrey-bion, is the current version of https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_configuration_properties clear enough or do you want me to make further updates?

**Comment by joffrey-bion** [comment]
Thanks for notifying me! I'm sorry, but I still don't see that level 4 in the table is not available to the Gradle wrapper. It says *"Stored in a gradle.properties file in the GRADLE_HOME, the optional Gradle installation directory."*.

In our past conversation, you said *"Gradle and the Wrapper don't use GRADLE_HOME."*, and then didn't clarify what you meant by this. Do you mean that it's just an environment variable name that people tend to use to point to a Gradle installation dir, but that nothing within Gradle or the wrapper scripts actually reads the `GRADLE_HOME` as a conventional name?

If this is the case, then this env variable should not be mentioned in the documentation. Especially not in the environment variables section, nor in the list of places where Gradle looks (because it doesn't look there). Instead, this 4th place should be described as it really is. For instance, "a `gradle.properties` file in the Gradle installation directory. When using the Gradle wrapper, the installation directory is abstracted away from the user, and there is usually no `gradle.properties` file there." (this is made up, maybe it's not how it works!).

**Comment by tknerr** [comment]
Hi @lkasso @joffrey-bion are there any news or further clarification regarding this?

I found this thread, because we stumbled over the same issue as @joffrey-bion, i.e. we are using gradle wrapper and want to set system wide "sane defaults" that can be overridden via project specific gradle.properties. Following the documentation in https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_configuration_properties I wonder what `$GRADLE_HOME` means for `gradlew` and where to place these system wide "sane defaults" for gradle properties.
